### DISCUSSION TITLE
Is it possible to add an input of the neural network which does not have a gradient in the ODEs?

### DISCUSSION THREAD

**Discussion by haowang1991** [lululxvi/deepxde#1334] [Q&A] [answered]
Hi,
I’m applying PINNs on dynamics and the ODEs I used are kinematics equations, such as:
dX/dt = Vx
dY/dt = Vy
dZ/dt = Vz
dVx/dt = ax + Ux (external input)
dVy/dt = ay + Uy (external input)
dVz/dt = az

From my understanding, in the framework, the input of the neural network is t = time, and the outputs are X, Y, Z, Vx, Vy, and Vz, because all the derivatives in the equations are time-derivative. However, in the real world, the dynamics is not time-dependent only.  
I wondered, is it possible to add other inputs of neural networks (such as positions and velocities at step n-1, external inputs) without changing the ODEs?

Thank you and look forward to comments!
HW

**Comment by haison19952013** [comment]
Yes, you can add any features that can affect the output. The physics information just works as regularization terms that make your NN avoid overfitting.

**Reply by haowang1991** [reply]
Thanks for your reply! I think the physics information is the loss function. In DeepXDE, it seems the inputs to the ODE/PDE are the SAME as the inputs to the neural network. Do you know how to change the inputs to the neural network but not changing the ODE/PDE?

**Reply by haison19952013** [reply, marked as answer]
Just expand the dimension of your inputs.It will not affect the way you define your ODE/PDE. In your case, assume that you have 2 inputs: `t` and additional control inputs `U` You can define the domain as below:
```
U_domain = dde.geometry.Interval(-1, 1)
timedomain = dde.geometry.TimeDomain(0, 1)
U_with_time_domain = dde.geometry.GeometryXTime(geom, timedomain)
```
For multi-dimensional inputs, you can refer to https://deepxde.readthedocs.io/en/latest/modules/deepxde.geometry.html#module-deepxde.geometry.geometry_nd. Remind that if you want to use `.Hypercube()`, you won't need `.GeometryXTime()` anymore. Keep in mind that, `.Interval()` or `.TimeDomain()` is just a convenient way to express your computation DOMAIN. In real cases, you can extend your domain dimensions as much as you want by using `.Hypercube()` .

**Reply by haowang1991** [reply]
Thank you so much! It really helps me and I'll definitely try that. Just one more question, shall I use domain = dde.deepxde.geometry.geometry_nd.Hypercube([t_min, x_min, y_min, z_min], [t_max, x_max, y_max, z_max])?
Since all my ODEs are time-derivative, I'm not sure if it is possible to remove time from the inputs, such as only use domain = dde.deepxde.geometry.geometry_nd.Hypercube([x_min, y_min, z_min], [x_max, y_max, z_max]).

**Reply by haison19952013** [reply]
- Question 1: It seems right; just check the document to make sure that. In addition, if you tend to use `.Hypercube()` only, you won't be able to use `.icbc.IC()` to define the initial conditions. In that scenario, I recommend you to `icbc.PointSetBC()`, which allows you to use the initial condition data/experimental data as additional constraints. You can refer to how to use `icbc.PointSetBC()` in [inverse problem here](https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse/lorenz.inverse.html)
- Question 2: You will need the `t`  input in the designed domain for calculating the physics-informed loss functions. Again, suppose you are working on a time-dependent problem; if you don't want to use `.GeometryXTime()`, you have to define all your desired inputs in `.Hypercube()`. By using `.Hypercube()`, you can control the index of the `t` input as you want, while if you use `.GeometryXTime()`, the index of `t` should be `-1`.
- Further recommendation: in case you have some difficulties with making your model predict the initial condition correctly, you can try to apply hard constraints to your model outputs. You can search with the keyword "hard constraint" in ISSUES, or you can refer to an example of hard constraint here: https://github.com/lululxvi/deepxde/issues/542. 
- Feel free to post a part of your code here so that the community can help you.

**Reply by haowang1991** [reply]
Thank you again. I'll try and post if I have more questions!
